grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy

```diff
@@ -314,15 +314,22 @@ class IndicatorDataService {
         Date date = LocalDate.now().minusMonths(querySize).toDate()
 
         def query = ReceiptItem.executeQuery("""
-            select s.shipmentNumber as number, s.name as name, count(ri.id) as count, s.requisition as requisition 
+            select 
+                s.shipmentNumber, 
+                s.name, 
+                count(ri.id), 
+                s.requisition
             from ReceiptItem as ri
+            left join ri.shipmentItem as si
             left join ri.receipt as r
             left join r.shipment as s
-            where r.receiptStatusCode = 'RECEIVED'
-            and s.destination = :location 
-            and ri.quantityShipped <> ri.quantityReceived
-            and r.actualDeliveryDate > :date 
+            where 
+                r.receiptStatusCode = 'RECEIVED'
+                and si.quantity <> ri.quantityReceived
```

Comment:
Can you explain why you kept this? This shouldn't matter and might cause some rows to be hidden.

Reply:
Using only `having sum(si.quantity) <> sum(ri.quantityReceived)` it counts each shipment's item from the moment it has at least one item with discrepancy:

For example, a shipment with 5 items, including 1 with a discrepancy, will  give a result of 5 instead of 1 in this case.

I wonder if the solution of our problem isn't somewhere else, putting `r.receiptStatusCode <> 'PENDING'` instead of `r.receiptStatusCode = 'RECEIVED'`.

Reply:
So the query you've written here will group by shipment, not by item, which means you're doing a giant sum across all shipment / receipt items and count of all receipt items. You need to have shipment item in the group by but since that gives you results in the wrong granularity (by item rather than by shipment) you need to do more processing.

You have two options. 
1) You either need two queries: One to get the item per row data (see my original query) and then another to get unique shipments with a count of discrepancies. This would be easier with straight SQL. Essentially use my query as an inner query and do the group by shipment / count shipment items around that. I'll add an example to the ticket.
2) Or you need to get the raw data in the query you've written (by adding shipment item id to the group by clause) and then do the group by shipment / count shipment items in code. I like this because it provides you with a way to include the detailed row data in your API response (for auditing purposes). I'll try to update the code to show you what I mean.

Reply:
Do we not want it grouped by shipment? I thought the indicator was meant to display per shipment the number of items that had been received with a discrepancy, not the total discrepancy per shipment.
If we group by item and then map it to shipment, we will get the total number of discrepancies, not the number of items.

Reply:
@bamoure I don't see a difference between what you said and what I did. I interpretted the requirements as saying return a data set with one row per shipment with a count of receipt discrepancies for that shipment.

The previous query was basically summing all quantities across all items in a shipment and then returning the count of all receipt items associated with the shipment ignoring the discrepancy at the line item level. 

| 'Number' |  'Name'                                                                      |  'Count' |  'Shipped Qty' |  'Received Qty' |  'Requisition ID'                   | 
|----------|------------------------------------------------------------------------------|----------|----------------|-----------------|-------------------------------------| 
| '068QBB' |  'Imres-DC-28Jan2020-CSNU6546848-CSNU6546848ImresPO4570+4843+4932+5011+5169' |  '142'   |  '2048191'     |  '2038041'      |  'ff8080816fa7b47c016fee3842a15792' | 
| '092AQM' |  'MissionpharmaA/S-DC-25Jun2019-MissionPharPO004238HaitiJAnBulk'             |  '3'     |  '21640'       |  '12800'        |  'ff8080816b4da7ca016b8fa7724604f8' | 
| '133QJU' |  'BostonHQ:HaitiStock-DC-01Mar2019-2019.3.1JPJoseph'                         |  '8'     |  '18'          |  '0'            |  '8a8a9e96691b644501693997eaf94a19' | 

That was after you removed this line:
```
and si.quantity <> ri.quantityReceived
```
The problem with keeping that line was that you basically went back to detecting discrepancies at the line item level. It should have been something like
```
and si.quantity <> sum(ri.quantityReceived)
```
But you can't do that in a single query (unless maybe you use having). The other approach would be to wrap the SQL that I wrote in the ticket in an outer query that groups by the shipment. I'll add an example there.

Reply:
Oh, right. Sorry I was getting a bit confused between the shipments and shipment items. Grouping by shipment after the query seems to work now!